### ISSUE TITLE
Async connect on Windows returns error when called before loop start

### ISSUE THREAD

**Issue by mcorino** [eclipse/mosquitto#373]
I am using mosquitto v1.4.10.

I found that although the documentation states that mosquitto_connect_async can "be called before or after mosquitto_loop_start" this does not hold on Windows (it does on Linux).
On Windows calling mosquitto_connect_async BEFORE calling mosquitto_connect_async results in error code MOSQ_ERR_ERRNO with errno set to 10057 (WSAENOTCONN) and no connection will be initialized. Calling mosquitto_connect_async after mosquitto_connect_async does work.

The following test code demonstrates the problem:

```c++
#include <iostream>
#include <string>
#include <cerrno>
#include <cstring>

#include <mosquitto.h>

#if defined(WIN32)
# include <windows.h>
# define sleep(x) Sleep((x)*1000)
#else
# include <unistd.h>
#endif

void handle_ret (int ret, const std::string& tag)
{
  std::cout << tag << std::endl;
  std::cout << mosquitto_strerror (ret) << std::endl;
  if (ret == MOSQ_ERR_ERRNO)
  {
    std::cout << errno << "=" << strerror (errno) << std::endl;
  }

  std::cout << std::endl;
}

int main (int , char* [])
{
  int ret;
  ret = mosquitto_lib_init ();
  handle_ret (ret, "init");

  struct mosquitto* c = mosquitto_new ("client-1", true, 0);

  ret = mosquitto_connect_async (c, "test.mosquitto.org", 1883, 60);
  handle_ret (ret, "connect_async");

  ret = mosquitto_loop_start (c);
  handle_ret (ret, "loop start");

  sleep (10);

  ret = mosquitto_disconnect (c);
  handle_ret (ret, "disconnect");

  ret = mosquitto_loop_stop (c, false);
  handle_ret (ret, "loop stop");

  ret = mosquitto_lib_cleanup ();
  handle_ret (ret, "clean up");

  return 0;
}
```

**Comment by mcorino** [comment]
Sorry, I somehow missed #365.
Consider this a detailing of that issue.

**Comment by ralight** [comment]
Thanks for the example code anyway, it's always helpful.

**Comment by mcorino** [comment]
You're welcome.

In the meantime I have more feedback on mosquitto on Windows.
I found that the async connect in general is not very reliable on Windows. When following the approach of calling async connect AFTER loop start at first everything seemed alright (no errors returned from the async connect call) and subsequent logging callbacks would (mostly) show a CONNECT being sent to the broker.
However the actual connection establishment turned out not to be too reliable. Sometimes the connection would be properly established and subsequent publish or subscribe calls would return successfully with actual publication and reception of topic instances but about just as often either the connection setup would stall at some point (meaning I did not ever see a CONNACK coming in, all publish calls would return a "no connection" error and stopping the connection and loop would somehow deadlock at the end) or I would see a MOSQ_ERR_ERRNO being returned somewhere with errno again set to 10057.
I found the only way to set up a connection that works reliable on both Windows and Linux is to use the sequence 'sync connect' followed by 'loop start'.

Probably important to note I tested all this against the mosquitto test broker at test.mosquitto.org.
The "less than perfect" network connection and performance reliability of that server was what I intended to test.
